python/dcr_template.py

```diff
@@ -132,78 +195,154 @@ def generate_templates_from_model(self, obsid_range=None, exposures=None, add_no
+
             if verbose:
-                print("\nFinished building template.")
-            # Weights may be different for each exposure
-            template[weights > 0] /= weights[weights > 0]
-            template[weights == 0] = 0.0
+                print(" ... Done!")
             if add_noise:
-                variance_level = np.median(calexp.getMaskedImage().getVariance().getArray())
                 rand_gen = np.random
-                template += rand_gen.normal(scale=np.sqrt(variance_level), size=template.shape)
+                template += rand_gen.normal(scale=np.sqrt(variance), size=template.shape)
 
-            dataId_out = self._build_dataId(obsid, self.photoParams.bandpass, instrument=instrument)[0]
-            exposure = self.create_exposure(template, variance=np.abs(template), snap=0,
-                                            elevation=el, azimuth=az, obsid=dataId_out['visit'])
+            if output_obsid_offset is not None:
+                obsid_out = obsid + output_obsid_offset
+            else:
+                obsid_out = obsid
+            dataId_out = self._build_dataId(obsid_out, self.filter_name, instrument=instrument)[0]
+            exposure = self.create_exposure(template, variance=variance, snap=0,
+                                            boresightRotAngle=rotation_angle,
+                                            elevation=el, azimuth=az, latitude=lat,
+                                            longitude=lon, altitude=alt, obsid=obsid_out)
             if warp:
                 wrap_warpExposure(exposure, wcs_exp, bbox_exp)
             if output_repository is not None:
+                if butler_out is None:
+                    butler_out = daf_persistence.Butler(output_repository)
                 butler_out.put(exposure, "calexp", dataId=dataId_out)
             yield exposure
 
+    def build_matched_template(self, exposure, model=None, el=None, rotation_angle=None, return_weights=True):
+        """Sub-routine to calculate the sum of the model images shifted by DCR for a given exposure.
+
+        Parameters
+        ----------
+        exposure : lsst.afw.image.ExposureD object
+            Single exposure to create a DCR-matched template for from the model.
+        model : List of numpy ndarrays, optional
+            The DCR model. If not set, then self.model is used.
+        el : lsst.afw.geom.Angle, optional
+            Elevation angle of the observation. If not set, it is read from the exposure.
+        rotation_angle : lsst.afw.geom.Angle, optional
+            Sky rotation angle of the observation. If not set it is calculated from the exposure metadata.
+        return_weights : bool, optional
+            Set to True to return the variance plane, as well as the image.
+
+        Returns
+        -------
+        Returns a numpy ndarray of the image values for the template.
+        If `return_weights` is set, then it returns a tuple of the image and variance arrays.
```

Comment:
I've added an explanation to the return value documentation. In short, this method was written to isolate the math and logic involved in extracting arrays of the image and weights values from a MaskedImage in an exposure, rather than duplicating the same steps in several places.